alerting/src/main/kotlin/org/opensearch/alerting/MonitorRunner.kt

```diff
@@ -757,39 +759,119 @@ object MonitorRunner : JobRunner, CoroutineScope, AbstractLifecycleComponent() {
             return
         }
 
+        val count: Int = lastRunContext["shards_count"] as Int
+        val updatedLastRunContext = lastRunContext.toMutableMap()
+        for (i: Int in 0 until count) {
+            val shard = i.toString()
+            val maxSeqNo: Long = getMaxSeqNo(index, shard)
+            updatedLastRunContext[shard] = maxSeqNo.toString()
+        }
+
+        val queryDocIds = mutableMapOf<DocLevelQuery, Set<String>>()
+        val docsToQueries = mutableMapOf<String, MutableList<String>>()
         for (query in queries) {
-            runForEachQuery(monitor, lastRunContext, index, query)
+            val matchingDocIds = runForEachQuery(monitor, lastRunContext, updatedLastRunContext, index, query)
+            queryDocIds[query] = matchingDocIds
+            matchingDocIds.forEach {
+                if (docsToQueries.containsKey(it)) {
+                    docsToQueries[it]?.add(query.id)
+                } else {
+                    docsToQueries[it] = mutableListOf(query.id)
+                }
+            }
         }
+
+        val queryIds = queries.map { it.id }
+
+        for (trigger in monitor.triggers) {
+            val triggerCtx = DocumentLevelTriggerExecutionContext(monitor, trigger as DocumentLevelTrigger)
+            val triggerResult = triggerService.runDocLevelTrigger(monitor, trigger, triggerCtx, docsToQueries, queryIds)
+
+            logger.info("trigger results")
+            logger.info(triggerResult.triggeredDocs.toString())
+
+            queryDocIds.forEach {
+                val queryTriggeredDocs = it.value.intersect(triggerResult.triggeredDocs)
+                if (queryTriggeredDocs.isNotEmpty()) {
+                    val findingId = createFindings(monitor, index, it.key, queryTriggeredDocs, trigger)
+                    // TODO: check if need to create alert, if so create it and point it to FindingId
+                    // TODO: run action as well, but this mat need to be throttled based on Mo's comment for bucket level alerting
+                }
+            }
```

Comment:
Abstract this out as a separate ``runForEachTrigger``. Again to achieve higher parallelism we can expect each trigger execution to return the findings. Here we may only want to consolidate them to create final findings. Thoughts?

Reply:
I have separated the function out for this. Also not sure if there is any benefit to consolidate all the findings in the end since we just want to run a single action based on each finding